File: CiaiControllerSDK/WebServer/HttpsOptions.cs
using System;

namespace CiaiControllerSDK.WebServer
{
    /// <summary>
    /// 服务器配置选项（支持HTTP和HTTPS）
    /// 参照 IncubatorController 的 Spring Boot SSL 配置
    /// </summary>
    public class HttpsOptions
    {
        /// <summary>
        /// 监听端口，默认443（HTTPS）或8080（HTTP）
        /// </summary>
        public int Port { get; set; } = 443;

        /// <summary>
        /// 是否启用HTTPS，默认true
        /// </summary>
        public bool UseHttps { get; set; } = true;

        /// <summary>
        /// 服务端证书文件路径（.pfx/.p12格式），HTTPS模式必需
        /// 对应 Spring Boot: server.ssl.key-store
        /// </summary>
        public string ServerCertificatePath { get; set; }

        /// <summary>
        /// 服务端证书密码
        /// 对应 Spring Boot: server.ssl.key-store-password
        /// </summary>
        public string ServerCertificatePassword { get; set; }

        /// <summary>
        /// 密钥别名
        /// 对应 Spring Boot: server.ssl.key-alias
        /// </summary>
        public string KeyAlias { get; set; }

        /// <summary>
        /// 密钥库类型，默认 PKCS12
        /// 对应 Spring Boot: server.ssl.key-store-type
        /// </summary>
        public string KeyStoreType { get; set; } = "PKCS12";

        /// <summary>
        /// 信任库文件路径，用于客户端证书验证
        /// 对应 Spring Boot: server.ssl.trust-store
        /// 如果不配置，将使用密钥库作为信任库（与 IncubatorController 一致）
        /// </summary>
        public string TrustStorePath { get; set; }

        /// <summary>
        /// 信任库密码
        /// 对应 Spring Boot: server.ssl.trust-store-password
        /// </summary>
        public string TrustStorePassword { get; set; }

        /// <summary>
        /// 信任库类型，默认 PKCS12
        /// </summary>
        public string TrustStoreType { get; set; } = "PKCS12";

        /// <summary>
        /// SSL/TLS 协议版本，默认 TLSv1.2（兼容 Java 8 基线）
        /// 对应 Spring Boot: server.ssl.protocol
        /// </summary>
        public string Protocol { get; set; } = "TLSv1.2";

        /// <summary>
        /// 启用的协议列表
        /// 对应 Spring Boot: server.ssl.enabled-protocols
        /// </summary>
        public string[] EnabledProtocols { get; set; } = new[] { "TLSv1.2" };

        /// <summary>
        /// 加密套件列表
        /// 对应 Spring Boot: server.ssl.ciphers
        /// </summary>
        public string[] Ciphers { get; set; } = Array.Empty<string>();

        /// <summary>
        /// 客户端认证模式
        /// 对应 Spring Boot: server.ssl.client-auth
        /// </summary>
        public ClientAuthMode ClientAuth { get; set; } = ClientAuthMode.Need;

        /// <summary>
        /// 是否要求客户端证书验证，默认true（仅HTTPS模式有效）
        /// </summary>
        public bool RequireClientCertificate { get; set; } = true;

        /// <summary>
        /// 受信任的客户端证书指纹列表
        /// </summary>
        public string[] TrustedClientThumbprints { get; set; }

        /// <summary>
        /// 受信任的CA证书指纹列表
        /// </summary>
        public string[] TrustedIssuerThumbprints { get; set; }

        /// <summary>
        /// 回调URL地址
        /// </summary>
        public string CallbackUrl { get; set; }

        /// <summary>
        /// 回调超时时间（毫秒），默认30000
        /// </summary>
        public int CallbackTimeoutMs { get; set; } = 30000;

        /// <summary>
        /// 是否启用回调，默认true（当CallbackUrl不为空时）
        /// </summary>
        public bool EnableCallback { get; set; } = true;

        /// <summary>
        /// 主机名，默认localhost
        /// </summary>
        public string Host { get; set; } = "localhost";

        public int MaxConcurrentRequests { get; set; } = 100;
        public int MaxRequestBodyBytes { get; set; } = 1024 * 1024;
        public int FunctionQueueCapacity { get; set; } = 100;
        public int IdempotencyCapacity { get; set; } = 10000;
        public int ShutdownTimeoutMs { get; set; } = 30000;

        /// <summary>允许SDK替换端口上已有的不同HTTPS证书绑定；默认禁止破坏性替换。</summary>
        public bool AllowReplaceCertificateBinding { get; set; }

        /// <summary>
        /// 获取完整的监听前缀（根据UseHttps返回http或https）
        /// </summary>
        public string GetListenPrefix()
        {
            var scheme = UseHttps ? "https" : "http";
            var listenerHost = Host?.Trim();
            if (string.IsNullOrWhiteSpace(listenerHost) || listenerHost == "0.0.0.0" ||
                listenerHost == "*" || listenerHost == "+")
            {
                listenerHost = "+";
            }

            return $"{scheme}://{listenerHost}:{Port}/";
        }

        /// <summary>
        /// 获取localhost监听前缀（备用，权限不足时使用）
        /// </summary>
        public string GetLocalhostPrefix()
        {
            var scheme = UseHttps ? "https" : "http";
            return $"{scheme}://localhost:{Port}/";
        }

        /// <summary>
        /// 验证配置是否有效
        /// </summary>
        public void Validate()
        {
            if (Port <= 0 || Port > 65535)
                throw new ArgumentException($"无效的端口号: {Port}");

            if (string.IsNullOrWhiteSpace(Host))
                throw new ArgumentException("监听主机名不能为空");
            if (MaxConcurrentRequests <= 0 || MaxRequestBodyBytes <= 0 ||
                FunctionQueueCapacity <= 0 || IdempotencyCapacity <= 0 || ShutdownTimeoutMs <= 0)
                throw new ArgumentException("服务器并发、请求体、队列、幂等记录和停机超时配置必须大于0");

            // HTTPS模式需要证书
            if (UseHttps)
            {
                if (string.IsNullOrEmpty(ServerCertificatePath))
                    throw new ArgumentException("HTTPS模式需要配置服务端证书路径(ServerCertificatePath)");

                if (!System.IO.File.Exists(ServerCertificatePath))
                    throw new ArgumentException($"服务端证书文件不存在: {ServerCertificatePath}");

                // 客户端证书验证配置检查
                // 如果 ClientAuth 为 Need 或 Want，则需要配置信任
                if (ClientAuth == ClientAuthMode.Need || ClientAuth == ClientAuthMode.Want)
                {
                    // 信任库路径可以为空，此时使用密钥库作为信任库（与 IncubatorController 一致）
                    // 如果配置了信任库路径，则验证文件是否存在
                    if (!string.IsNullOrEmpty(TrustStorePath) && !System.IO.File.Exists(TrustStorePath))
                    {
                        throw new ArgumentException($"信任库文件不存在: {TrustStorePath}");
                    }
                }
            }
        }

        /// <summary>
        /// 创建HTTP模式配置（无需证书）
        /// </summary>
        public static HttpsOptions CreateHttp(int port = 8080, string host = "localhost")
        {
            return new HttpsOptions
            {
                UseHttps = false,
                Port = port,
                Host = host,
                RequireClientCertificate = false,
                ClientAuth = ClientAuthMode.None
            };
        }

        /// <summary>
        /// 创建HTTPS模式配置
        /// </summary>
        public static HttpsOptions CreateHttps(int port = 443, string host = "localhost",
            string certPath = null, string certPassword = null)
        {
            return new HttpsOptions
            {
                UseHttps = true,
                Port = port,
                Host = host,
                ServerCertificatePath = certPath,
                ServerCertificatePassword = certPassword
            };
        }

        /// <summary>
        /// 创建完整的HTTPS配置（参照 IncubatorController 的 Spring Boot 配置）
        /// </summary>
        /// <param name="port">端口</param>
        /// <param name="host">主机</param>
        /// <param name="keyStorePath">密钥库路径 (server.ssl.key-store)</param>
        /// <param name="keyStorePassword">密钥库密码 (server.ssl.key-store-password)</param>
        /// <param name="keyStoreType">密钥库类型 (server.ssl.key-store-type)</param>
        /// <param name="keyAlias">密钥别名 (server.ssl.key-alias)</param>
        /// <param name="trustStorePath">信任库路径 (server.ssl.trust-store)，可为null则使用密钥库</param>
        /// <param name="trustStorePassword">信任库密码 (server.ssl.trust-store-password)</param>
        /// <param name="protocol">SSL协议 (server.ssl.protocol)</param>
        /// <param name="ciphers">加密套件 (server.ssl.ciphers)</param>
        /// <param name="clientAuth">客户端认证模式 (server.ssl.client-auth)</param>
        public static HttpsOptions CreateHttpsFull(
            int port, string host,
            string keyStorePath, string keyStorePassword, string keyStoreType, string keyAlias,
            string trustStorePath, string trustStorePassword,
            string protocol, string[] ciphers,
            ClientAuthMode clientAuth)
        {
            return new HttpsOptions
            {
                Port = port,
                Host = host,
                UseHttps = true,

                // 密钥库配置
                ServerCertificatePath = keyStorePath,
                ServerCertificatePassword = keyStorePassword,
                KeyStoreType = keyStoreType ?? "PKCS12",
                KeyAlias = keyAlias,

                // 信任库配置
                TrustStorePath = trustStorePath,
                TrustStorePassword = trustStorePassword,

                // TLS配置
                Protocol = protocol ?? "TLSv1.2",
                EnabledProtocols = new[] { protocol ?? "TLSv1.2" },
                Ciphers = ciphers ?? Array.Empty<string>(),

                // 客户端认证
                ClientAuth = clientAuth,
                RequireClientCertificate = clientAuth != ClientAuthMode.None
            };
        }
    }

    /// <summary>
    /// 客户端认证模式
    /// 对应 Spring Boot: server.ssl.client-auth
    /// </summary>
    public enum ClientAuthMode
    {
        /// <summary>
        /// 需要客户端证书
        /// </summary>
        Need,

        /// <summary>
        /// 想要客户端证书，但不是必须
        /// </summary>
        Want,

        /// <summary>
        /// 不需要客户端证书
        /// </summary>
        None
    }
}

File: CiaiControllerSDKForJava/src/main/java/com/ciai/controller/sdk/webserver/HttpsOptions.java
package com.ciai.controller.sdk.webserver;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * HTTPS服务器配置选项
 * 参照 IncubatorController 的 Spring Boot SSL 配置实现
 */
public class HttpsOptions {

    // 基本服务器配置
    private int port = 443;
    private String host = "localhost";
    private boolean useHttps = true;

    // 服务端证书配置 (对应 Spring Boot: server.ssl.key-store-*)
    private String serverCertificatePath;
    private String serverCertificatePassword;
    private String keyAlias;  // 密钥别名 (对应 Spring Boot: server.ssl.key-alias)
    private String keyStoreType = "PKCS12";  // 密钥库类型 (对应 Spring Boot: server.ssl.key-store-type)

    // 信任库配置 (对应 Spring Boot: server.ssl.trust-store-*)
    private String trustStorePath;
    private String trustStorePassword;
    private String trustStoreType = "PKCS12";  // 信任库类型

    // TLS 协议配置 (对应 Spring Boot: server.ssl.protocol, server.ssl.enabled-protocols)
    private String protocol = "TLSv1.2";  // Java 8与现代运行时共同支持的安全基线
    private String[] enabledProtocols = new String[]{"TLSv1.2"};  // 启用的协议

    // 加密套件配置 (对应 Spring Boot: server.ssl.ciphers)
    private String[] ciphers = new String[0]; // 空数组表示使用JVM安全策略

    // 客户端认证配置 (对应 Spring Boot: server.ssl.client-auth)
    private ClientAuthMode clientAuth = ClientAuthMode.NEED;  // need, want, none
    private boolean requireClientCertificate = true;
    private String[] trustedClientThumbprints = new String[0];
    private String[] trustedIssuerThumbprints = new String[0];

    // 回调配置
    private String callbackUrl;
    private int callbackTimeoutMs = 30000;
    private boolean enableCallback = true;
    private int maxConcurrentRequests = 100;
    private int maxRequestBodyBytes = 1024 * 1024;
    private int functionQueueCapacity = 100;
    private int idempotencyCapacity = 10000;
    private int shutdownTimeoutMs = 30000;

    /**
     * 客户端认证模式
     * 对应 Spring Boot: server.ssl.client-auth
     */
    public enum ClientAuthMode {
        NEED,    // 需要客户端证书
        WANT,    // 想要客户端证书，但不是必须
        NONE     // 不需要客户端证书
    }

    public HttpsOptions() {
    }

    // Getters and Setters
    public int getPort() {
        return port;
    }

    public void setPort(int port) {
        this.port = port;
    }

    public String getHost() {
        return host;
    }

    public void setHost(String host) {
        this.host = host;
    }

    public boolean isUseHttps() {
        return useHttps;
    }

    public void setUseHttps(boolean useHttps) {
        this.useHttps = useHttps;
    }

    public String getServerCertificatePath() {
        return serverCertificatePath;
    }

    public void setServerCertificatePath(String serverCertificatePath) {
        this.serverCertificatePath = serverCertificatePath;
    }

    public String getServerCertificatePassword() {
        return serverCertificatePassword;
    }

    public void setServerCertificatePassword(String serverCertificatePassword) {
        this.serverCertificatePassword = serverCertificatePassword;
    }

    public boolean isRequireClientCertificate() {
        return requireClientCertificate;
    }

    public void setRequireClientCertificate(boolean requireClientCertificate) {
        this.requireClientCertificate = requireClientCertificate;
    }

    public String[] getTrustedClientThumbprints() {
        return trustedClientThumbprints;
    }

    public void setTrustedClientThumbprints(String[] trustedClientThumbprints) {
        this.trustedClientThumbprints = trustedClientThumbprints;
    }

    public String[] getTrustedIssuerThumbprints() {
        return trustedIssuerThumbprints;
    }

    public void setTrustedIssuerThumbprints(String[] trustedIssuerThumbprints) {
        this.trustedIssuerThumbprints = trustedIssuerThumbprints;
    }

    public String getCallbackUrl() {
        return callbackUrl;
    }

    public void setCallbackUrl(String callbackUrl) {
        this.callbackUrl = callbackUrl;
    }

    public int getCallbackTimeoutMs() {
        return callbackTimeoutMs;
    }

    public void setCallbackTimeoutMs(int callbackTimeoutMs) {
        this.callbackTimeoutMs = callbackTimeoutMs;
    }

    public boolean isEnableCallback() {
        return enableCallback;
    }

    public void setEnableCallback(boolean enableCallback) {
        this.enableCallback = enableCallback;
    }

    public int getMaxConcurrentRequests() { return maxConcurrentRequests; }
    public void setMaxConcurrentRequests(int value) { this.maxConcurrentRequests = value; }
    public int getMaxRequestBodyBytes() { return maxRequestBodyBytes; }
    public void setMaxRequestBodyBytes(int value) { this.maxRequestBodyBytes = value; }
    public int getFunctionQueueCapacity() { return functionQueueCapacity; }
    public void setFunctionQueueCapacity(int value) { this.functionQueueCapacity = value; }
    public int getIdempotencyCapacity() { return idempotencyCapacity; }
    public void setIdempotencyCapacity(int value) { this.idempotencyCapacity = value; }
    public int getShutdownTimeoutMs() { return shutdownTimeoutMs; }
    public void setShutdownTimeoutMs(int value) { this.shutdownTimeoutMs = value; }

    // ========== 新增的配置项 Getter/Setter ==========

    public String getKeyAlias() {
        return keyAlias;
    }

    public void setKeyAlias(String keyAlias) {
        this.keyAlias = keyAlias;
    }

    public String getKeyStoreType() {
        return keyStoreType;
    }

    public void setKeyStoreType(String keyStoreType) {
        this.keyStoreType = keyStoreType;
    }

    public String getTrustStorePath() {
        return trustStorePath;
    }

    public void setTrustStorePath(String trustStorePath) {
        this.trustStorePath = trustStorePath;
    }

    public String getTrustStorePassword() {
        return trustStorePassword;
    }

    public void setTrustStorePassword(String trustStorePassword) {
        this.trustStorePassword = trustStorePassword;
    }

    public String getTrustStoreType() {
        return trustStoreType;
    }

    public void setTrustStoreType(String trustStoreType) {
        this.trustStoreType = trustStoreType;
    }

    public String getProtocol() {
        return protocol;
    }

    public void setProtocol(String protocol) {
        this.protocol = protocol;
    }

    public String[] getEnabledProtocols() {
        return enabledProtocols;
    }

    public void setEnabledProtocols(String[] enabledProtocols) {
        this.enabledProtocols = enabledProtocols;
    }

    public String[] getCiphers() {
        return ciphers;
    }

    public void setCiphers(String[] ciphers) {
        this.ciphers = ciphers;
    }

    public ClientAuthMode getClientAuth() {
        return clientAuth;
    }

    public void setClientAuth(ClientAuthMode clientAuth) {
        this.clientAuth = clientAuth;
        // 同步更新 requireClientCertificate
        this.requireClientCertificate = (clientAuth == ClientAuthMode.NEED || clientAuth == ClientAuthMode.WANT);
    }

    /**
     * 获取监听前缀
     */
    public String getListenPrefix() {
        String scheme = useHttps ? "https" : "http";
        return scheme + "://" + host + ":" + port + "/";
    }

    /**
     * 验证配置
     */
    public void validate() {
        List<String> errors = new ArrayList<>();

        if (port <= 0 || port > 65535) {
            errors.add("Invalid port: " + port);
        }
        if (maxConcurrentRequests <= 0 || maxRequestBodyBytes <= 0
                || functionQueueCapacity <= 0 || idempotencyCapacity <= 0
                || shutdownTimeoutMs <= 0) {
            errors.add("Server concurrency, body, queue, idempotency and shutdown limits must be positive");
        }

        if (useHttps) {
            if (serverCertificatePath == null || serverCertificatePath.isEmpty()) {
                errors.add("Server certificate path (key-store) is required for HTTPS");
            }

            // 如果需要客户端证书验证，检查信任库配置
            if (clientAuth == ClientAuthMode.NEED || clientAuth == ClientAuthMode.WANT) {
                // 信任库路径可以为空，此时使用密钥库作为信任库（与 IncubatorController 一致）
                // 但如果配置了信任库路径，则必须有密码
                if (trustStorePath != null && !trustStorePath.isEmpty()) {
                    // 信任库已配置，OK
                }
                // 如果没有配置信任库，则使用密钥库作为信任库（与 IncubatorController 的做法一致）
            }
        }

        if (!errors.isEmpty()) {
            throw new IllegalArgumentException("Configuration errors: " + String.join(", ", errors));
        }
    }

    /**
     * 创建HTTP配置
     */
    public static HttpsOptions createHttp(int port) {
        return createHttp(port, "localhost");
    }

    /**
     * 创建HTTP配置
     */
    public static HttpsOptions createHttp(int port, String host) {
        HttpsOptions options = new HttpsOptions();
        options.setPort(port);
        options.setHost(host);
        options.setUseHttps(false);
        options.setRequireClientCertificate(false);
        return options;
    }

    /**
     * 创建HTTPS配置
     */
    public static HttpsOptions createHttps(int port, String host, String certPath, String certPassword) {
        HttpsOptions options = new HttpsOptions();
        options.setPort(port);
        options.setHost(host);
        options.setUseHttps(true);
        options.setServerCertificatePath(certPath);
        options.setServerCertificatePassword(certPassword);
        return options;
    }

    /**
     * 创建完整的HTTPS配置（参照 IncubatorController 的 Spring Boot 配置）
     * @param port 端口
     * @param host 主机
     * @param keyStorePath 密钥库路径 (server.ssl.key-store)
     * @param keyStorePassword 密钥库密码 (server.ssl.key-store-password)
     * @param keyStoreType 密钥库类型 (server.ssl.key-store-type)
     * @param keyAlias 密钥别名 (server.ssl.key-alias)
     * @param trustStorePath 信任库路径 (server.ssl.trust-store)，可为null则使用密钥库
     * @param trustStorePassword 信任库密码 (server.ssl.trust-store-password)
     * @param protocol SSL协议 (server.ssl.protocol)
     * @param ciphers 加密套件 (server.ssl.ciphers)
     * @param clientAuth 客户端认证模式 (server.ssl.client-auth)
     */
    public static HttpsOptions createHttpsFull(
            int port, String host,
            String keyStorePath, String keyStorePassword, String keyStoreType, String keyAlias,
            String trustStorePath, String trustStorePassword,
            String protocol, String[] ciphers,
            ClientAuthMode clientAuth) {
        HttpsOptions options = new HttpsOptions();
        options.setPort(port);
        options.setHost(host);
        options.setUseHttps(true);

        // 密钥库配置
        options.setServerCertificatePath(keyStorePath);
        options.setServerCertificatePassword(keyStorePassword);
        options.setKeyStoreType(keyStoreType != null ? keyStoreType : "PKCS12");
        options.setKeyAlias(keyAlias);

        // 信任库配置
        options.setTrustStorePath(trustStorePath);
        options.setTrustStorePassword(trustStorePassword);

        // TLS配置
        options.setProtocol(protocol != null ? protocol : "TLSv1.2");
        options.setEnabledProtocols(new String[]{options.getProtocol()});
        options.setCiphers(ciphers != null ? ciphers : new String[0]);

        // 客户端认证
        options.setClientAuth(clientAuth != null ? clientAuth : ClientAuthMode.NEED);

        return options;
    }

    @Override
    public String toString() {
        return "HttpsOptions{" +
                "port=" + port +
                ", host='" + host + '\'' +
                ", useHttps=" + useHttps +
                ", keyStoreType='" + keyStoreType + '\'' +
                ", keyAlias='" + keyAlias + '\'' +
                ", protocol='" + protocol + '\'' +
                ", enabledProtocols=" + Arrays.toString(enabledProtocols) +
                ", ciphers=" + Arrays.toString(ciphers) +
                ", clientAuth=" + clientAuth +
                ", trustStorePath='" + trustStorePath + '\'' +
                ", callbackUrl='" + callbackUrl + '\'' +
                '}';
    }
}
